src/serde_utils.rs

```diff
@@ -0,0 +1,107 @@
+
+//! Module for special serde serializations.
+
+pub mod btreemap {
+    //! Module for serialization of BTreeMaps because serde_json will
+    //! not serialize hashmaps with non-string keys be default.
+    #![allow(missing_docs)]
+
+    // NOTE: This module can be exactly copied to use with HashMap.
+
+    use std::collections::BTreeMap;
+    use serde;
+
+    pub fn serialize<S, T, U>(v: &BTreeMap<T, U>, s: S)
+        -> Result<S::Ok, S::Error> where
+        S: serde::Serializer,
+        T: serde::Serialize + ::std::hash::Hash + Eq + Ord,
+        U: serde::Serialize,
+    {
+        use serde::ser::SerializeSeq;
+
+        let mut seq = s.serialize_seq(Some(v.len()))?;
+        for pair in v.iter() {
+            seq.serialize_element(&pair)?;
+        }
+        seq.end()
```

Comment:
What's the reason for serializing a sequence of pairs instead of using the map interfaces?

Reply:
Hmm, this was just taken as is from another impl (either from the macro or from one I have in another crate). I didn't know there was an API for maps. Let me look into those.

Reply:
Oh.. The entire point of that module is to not use the map impl because JSON only supports that when the keys are strings and we don't want to have to use to_string and allocate strings. Let me see if there's a way to do this using serde right now. (Since I already have this attempted.) I think what we're missing is some way to hand serde the object not as a serializable but explicitly saying it must become a string or something (like what serialize_str does with types that have Display). It might be missing that one on the Map serializers.

Reply:
So yeah the whole point of this module is to work around the fact that serde_json needs keys to be strings and we don't want to allocate strings for all keys.

Reply:
And the effective overhead is two characters per map entry, so that's quite fine. I think that's better than one allocation per map entry.

But if you want to give it a shot and try to work with the API to have serde_json accept the map keys, here is what you'd start with:
```diff
diff --cc src/serde_utils.rs
index 0c0b243,0c0b243..debc771
--- a/src/serde_utils.rs
+++ b/src/serde_utils.rs
@@@ -17,13 -17,13 +17,13 @@@ pub mod btreemap 
          T: serde::Serialize + ::std::hash::Hash + Eq + Ord,
          U: serde::Serialize,
      {
--        use serde::ser::SerializeSeq;
++        use serde::ser::SerializeMap;
  
--        let mut seq = s.serialize_seq(Some(v.len()))?;
--        for pair in v.iter() {
--            seq.serialize_element(&pair)?;
++        let mut map = s.serialize_map(Some(v.len()))?;
++        for (key, value) in v.iter() {
++            map.serialize_entry(key, value)?;
          }
--        seq.end()
++        map.end()
      }
  
      pub fn deserialize<'de, D, T, U>(d: D)
@@@ -42,14 -42,14 +42,14 @@@
              type Value = BTreeMap<T, U>;
  
              fn expecting(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
--                write!(f, "a sequence of pairs")
++                write!(f, "a map with string keys")
              }
  
--            fn visit_seq<A: serde::de::SeqAccess<'de>>(self, mut a: A)
++            fn visit_map<A: serde::de::MapAccess<'de>>(self, mut a: A)
                  -> Result<Self::Value, A::Error>
              {
                  let mut ret = BTreeMap::new();
--                while let Some((key, value)) = a.next_element()? {
++                while let Some((key, value)) = a.next_entry()? {
                      ret.insert(key, value);
                  }
                  Ok(ret)
```

Reply:
Ah, I see, the key probably couldn't be an object.